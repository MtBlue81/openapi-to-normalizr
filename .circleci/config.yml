version: 2
base_image: &base_image
  image: circleci/node:10.15.1

jobs:
  npm_install:
    docker:
      - <<: *base_image
    steps:
      - checkout
      - restore_cache:
          name: Restore Yarn Package Cache
          keys:
            - yarn-packages-v2-{{ .Branch }}-{{ checksum "yarn.lock" }}
            - yarn-packages-v2-{{ .Branch }}
            - yarn-packages-v2-master
            - yarn-packages-v2-
      - run:
          name: Install Dependencies
          command: yarn install --frozen-lockfile
      - save_cache:
          name: Save Yarn Package Cache
          key: yarn-packages-v2-{{ .Branch }}-{{ checksum "yarn.lock" }}
          paths:
            - node_modules/
  lint:
    docker:
      - <<: *base_image
    steps:
      - checkout
      - restore_cache:
          name: Restore Yarn Package Cache
          keys:
            - yarn-packages-v2-{{ .Branch }}-{{ checksum "yarn.lock" }}
            - yarn-packages-v2-{{ .Branch }}
            - yarn-packages-v2-master
            - yarn-packages-v2-
      - run:
          name: Spec lint
          command: yarn speclint
  test:
    docker:
      - <<: *base_image
    steps:
      - checkout
      - restore_cache:
          name: Restore Yarn Package Cache
          keys:
            - yarn-packages-v2-{{ .Branch }}-{{ checksum "yarn.lock" }}
            - yarn-packages-v2-{{ .Branch }}
            - yarn-packages-v2-master
            - yarn-packages-v2-
      - run:
          name: Test
          command: yarn test
workflows:
  version: 2
  test:
    jobs:
      - npm_install
      - lint:
          requires:
            - npm_install
      - test:
          requires:
            - npm_install
