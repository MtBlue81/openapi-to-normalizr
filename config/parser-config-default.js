const path = require('path');
const templates = {
  model: 'templates/model_template.mustache',
  override: 'templates/override_template.mustache',
  schema: 'templates/schema_template.mustache',
  head: 'templates/head_template.mustache',
  dependency: 'templates/dependency_template.mustache',
  models: 'templates/models_template.mustache',
  actionTypes: 'templates/action_types_template.mustache',
  spec: 'templates/spec_template.mustache',
  oneOf: 'templates/one_of_template.mustache',
};
module.exports = {
  templates: Object.keys(templates).reduce((acc, key) => {
    acc[key] = path.join(__dirname, '../', templates[key]);
    return acc;
  }, {}),
  modelsDir: 'tmp',
  outputDir: 'tmp',
  useFlow: true,
  usePropType: true,
};
