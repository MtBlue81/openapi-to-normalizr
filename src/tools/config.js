class Config {
  constructor(config) {
    this._config = config;
    this.attributeConverter = config.attributeConverter ? config.attributeConverter : (str) => str;
    this.modelsDir = config.modelsDir || 'dist';
  }

  get tags() {
    return this._config.tags;
  }

  get outputPath() {
    return this._config.outputPath;
  }

  get useTypeScript() {
    return this._config.useTypeScript;
  }

  get extension() {
    return this._config.useTypeScript ? 'ts' : 'js';
  }

  formatForModelGenerator() {
    const {
      modelsDir: outputDir,
      templates: templatePath,
      usePropType,
      useFlow,
      useTypeScript,
      attributeConverter,
    } = this._config;
    return {
      outputDir,
      templatePath,
      usePropType,
      useFlow,
      useTypeScript,
      attributeConverter,
      extension: this.extension,
    };
  }

  formatForActionTypesGenerator() {
    const { outputPath, templates: templatePath, useTypeScript } = this._config;
    return {
      templatePath,
      outputPath: outputPath.actions,
      schemasFilePath: outputPath.schemas,
      useTypeScript,
      extension: this.extension,
    };
  }

  formatForSchemaGenerator() {
    const {
      outputPath,
      templates: templatePath,
      modelsDir,
      attributeConverter,
      useTypeScript,
    } = this._config;
    return {
      templatePath,
      outputPath: outputPath.schemas,
      modelsDir,
      attributeConverter,
      useTypeScript,
      extension: this.extension,
    };
  }

  formatForJsSpecGenerator() {
    const { templates: templatePath, outputPath } = this._config;
    return {
      templatePath,
      outputPath: outputPath.jsSpec,
      extension: this.extension,
    };
  }
}

module.exports = Config;
