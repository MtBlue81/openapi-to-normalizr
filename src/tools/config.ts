export default class Config {
  _config: ConfigObject;

  attributeConverter: AttributeConverter;

  modelsDir: string;

  constructor(config: ConfigObject) {
    this._config = config;
    this.attributeConverter = config.attributeConverter ? config.attributeConverter : (str) => str;
    this.modelsDir = config.modelsDir || 'dist';
  }

  get tags() {
    return this._config.tags;
  }

  get outputPath() {
    return this._config.outputPath;
  }

  get useTypeScript() {
    return this._config.useTypeScript;
  }

  get extension() {
    return this._config.useTypeScript ? 'ts' : 'js';
  }

  formatForModelGenerator() {
    const { templates: templatePath, usePropType, useTypeScript } = this._config;
    return {
      outputDir: this.modelsDir,
      templatePath,
      usePropType,
      useTypeScript,
      attributeConverter: this.attributeConverter,
      extension: this.extension,
    } as const;
  }

  formatForActionTypesGenerator() {
    const { outputPath, templates: templatePath, useTypeScript } = this._config;
    const operationIdList: string[] = [];
    return {
      outputPath: outputPath.actions,
      schemasFilePath: outputPath.schemas,
      templatePath,
      operationIdList,
      useTypeScript,
      extension: this.extension,
    } as const;
  }

  formatForSchemaGenerator() {
    const { outputPath, templates: templatePath, useTypeScript } = this._config;
    return {
      templatePath,
      outputPath: outputPath.schemas,
      modelsDir: this.modelsDir,
      attributeConverter: this.attributeConverter,
      useTypeScript,
      extension: this.extension,
    } as const;
  }

  formatForJsSpecGenerator() {
    const { templates: templatePath, outputPath } = this._config;
    return {
      templatePath,
      outputPath: outputPath.jsSpec,
      extension: this.extension,
    } as const;
  }
}
