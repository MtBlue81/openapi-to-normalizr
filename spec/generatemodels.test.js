import config from '../config/parser-config-default';
import rimraf from 'rimraf';
import { execSync } from 'child_process';
import fs from 'fs';
import path from 'path';

const outputDir = config.modelsDir;
const dir = __dirname;

const walk = (p, fileCallback) => {
  const files = fs.readdirSync(p);
  files.forEach((f) => {
    const fp = path.join(p, f);
    if (fs.statSync(fp).isDirectory()) {
      walk(fp, fileCallback);
    } else {
      fileCallback(fp);
    }
  });
};

const snapshot = (file) => {
  execSync(`${dir}/../bin/generatemodels --config ${dir}/../config/parser-config-default.js ${dir}/${file}`);
  walk(outputDir, (path) => {
    const output = fs.readFileSync(path, 'utf8');
    expect({ path, output }).toMatchSnapshot();
  });
}

describe('model generator spec', () => {
  beforeEach(() => new Promise((resolve) => rimraf(outputDir, resolve)));

  test('from json schema ref', () => {
    snapshot('json_schema_ref.yml')
  });

  test('parse oneOf type', () => {
    snapshot('one_of.yml')
  });
});
