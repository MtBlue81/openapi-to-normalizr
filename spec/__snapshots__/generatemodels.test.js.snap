// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`model generator spec from json schema ref 1`] = `
Object {
  "output": "/* eslint-disable */
// @flow
/**
 * generated from API definition file
 */

import { Record, List } from 'immutable';
import { schema as _schema, denormalize as _denormalize } from 'normalizr';
import isArray from 'lodash/isArray';

export const GENDER_MALE: string = 'male';
export const GENDER_FEMALE: string = 'female';
export const GENDER_OTHER: string = 'other';

const defaultValues = {
  name: undefined,
  gender: undefined,
};

export const schema = new _schema.Entity('Owner', {}, {idAttribute: 'name'});


/**
 * @params ids : Owner's id[s]
 * @params entities : all entities that need to denormalize ids
 */
export const denormalize = (ids: number[] | string[], entities: any) => _denormalize(ids, isArray(ids) || List.isList(ids) ? [schema] : schema, entities);

export default class Owner extends Record(defaultValues) {
  name: string;
  gender: string;
  static denormalize(id: number[] | string[], entities: any) {
    return denormalize(id, entities);
  }
}
",
  "path": "tmp/base/owner.js",
}
`;

exports[`model generator spec from json schema ref 2`] = `
Object {
  "output": "/* eslint-disable */
// @flow
/**
 * generated from API definition file
 */

import { Record, List } from 'immutable';
import { schema as _schema, denormalize as _denormalize } from 'normalizr';
import isArray from 'lodash/isArray';

export const KIND_DOG: string = 'Dog';
export const KIND_CAT: string = 'Cat';

const defaultValues = {
  id: undefined,
  kind: undefined,
};

export const schema = new _schema.Entity('Pet', {}, {idAttribute: 'id'});


/**
 * @params ids : Pet's id[s]
 * @params entities : all entities that need to denormalize ids
 */
export const denormalize = (ids: number[] | string[], entities: any) => _denormalize(ids, isArray(ids) || List.isList(ids) ? [schema] : schema, entities);

export default class Pet extends Record(defaultValues) {
  id: number;
  kind: string;
  static denormalize(id: number[] | string[], entities: any) {
    return denormalize(id, entities);
  }
}
",
  "path": "tmp/base/pet.js",
}
`;

exports[`model generator spec from json schema ref 3`] = `
Object {
  "output": "/* eslint-disable */
// @flow
/**
 * generated from API definition file
 */

import { Record, List } from 'immutable';
import { schema as _schema, denormalize as _denormalize } from 'normalizr';
import isArray from 'lodash/isArray';
import Owner, { schema as OwnerSchema } from './owner';

export const KIND_DOG: string = 'Dog';
export const KIND_CAT: string = 'Cat';

const defaultValues = {
  id: undefined,
  name: undefined,
  kind: undefined,
  owner: undefined,
};

export const schema = new _schema.Entity('WrappedPet', {}, {idAttribute: 'id'});
schema.define({
  owner: OwnerSchema
});


/**
 * @params ids : WrappedPet's id[s]
 * @params entities : all entities that need to denormalize ids
 */
export const denormalize = (ids: number[] | string[], entities: any) => _denormalize(ids, isArray(ids) || List.isList(ids) ? [schema] : schema, entities);

export default class WrappedPet extends Record(defaultValues) {
  id: number;
  name: string;
  kind: string;
  owner: Owner;
  static denormalize(id: number[] | string[], entities: any) {
    return denormalize(id, entities);
  }
}
",
  "path": "tmp/base/wrapped_pet.js",
}
`;

exports[`model generator spec from json schema ref 4`] = `
Object {
  "output": "/**
 * generated from API definition file
 */

import WrappedPet from './wrapped_pet';
import Pet from './pet';
import Owner from './owner';

export {
  WrappedPet,
  Pet,
  Owner,
};
",
  "path": "tmp/index.js",
}
`;

exports[`model generator spec from json schema ref 5`] = `
Object {
  "output": "/* eslint-disable comma-dangle */
/**
 * generated from API definition file
 */

import _Owner from './base/owner';
export * from './base/owner';

export default class Owner extends _Owner {

  /**
   * write custom methods here
   */
}
",
  "path": "tmp/owner.js",
}
`;

exports[`model generator spec from json schema ref 6`] = `
Object {
  "output": "/* eslint-disable comma-dangle */
/**
 * generated from API definition file
 */

import _Pet from './base/pet';
export * from './base/pet';

export default class Pet extends _Pet {

  /**
   * write custom methods here
   */
}
",
  "path": "tmp/pet.js",
}
`;

exports[`model generator spec from json schema ref 7`] = `
Object {
  "output": "/* eslint-disable comma-dangle */
/**
 * generated from API definition file
 */

import _WrappedPet from './base/wrapped_pet';
export * from './base/wrapped_pet';

export default class WrappedPet extends _WrappedPet {

  /**
   * write custom methods here
   */
}
",
  "path": "tmp/wrapped_pet.js",
}
`;
